### PR TITLE
Implement calculation of population results

### PR DESCRIPTION
**Pull request by mgramigna**
# Summary

This PR adds functionality to the CLI for using the `$cql` operation of cqf-ruler to get calculation results for "Initial Population", "Numerator", and "Denominator"

**NOTE**: I will publish this to npm as a globally installed module after this gets merged in.

~**NOTE 2**: I am investigating some weird async behavior where the code continue before cqf-ruler fully processes the bundle. But review can still be done for a small # of patients.~

## New behavior

* Can now run the CLI providing a path to FHIR Patient Bundles, a `.cql` file, and a url for a running `cqf-ruler` instance to get a `.csv` of the results as well as the bundles sorted into directories based on the calculation results

## Code changes

* CLI behavior in `src/cli.js`
* HTTP requests and response parsing in `src/utils/calculation.js`
* Unit test for calculation in `test/calculation.test.js`
* JavaScript linting config in `.eslintrc.json` (deriving from [AirBnB style guide](https://github.com/airbnb/javascript)
* CI config in `.travis.yml` - Runs the linter and unit tests

# Testing guidance

## Functionality Testing

See the `README` for pre-reqs and more detailed options. The following is an example workflow for testing:

1) Clone this library and install dependencies (see README if needed)
2) Run cqf-ruler somewhere (e.g. `docker run -p 8080:8080 contentgroup/cqf-ruler`
3) Generate some `stu3` patients with Synthea (more detail [here](https://github.com/synthetichealth/synthea#developer-quick-start)): `./run_synthea -p 2`
4) Use the cli to get calculation results

``` bash
node src/cli.js -d /path/to/synthea/stu3/output/ -c /path/to/cql/file -u http://localhost:8080/cqf-ruler-dstu3/fhir
``` 

5) Observe output in `./output`

## Unit Testing

I have set up [jest](https://jestjs.io/) to make assertions about calculation under a controlled environment, and [nock](https://github.com/nock/nock) to mock HTTP responses for the `$cql` operation for unit testing. To run the unit tests, run

``` bash
npm test
```

### CODE REVIEW

**Comment by mgramigna** [comment]
For testing, this is the cql file I used. It's rather complex:

```
define "Initial Population":
    true

define "Numerator":
    true

define "Denominator":
    true
```

**Comment by mgramigna** [comment]
Pushed up a change that should properly halt until each bundle is posted

**Comment by mgramigna** [comment]
> I ran this:
> 
> ```shell
> node src/cli.js -d ../synthea/output/fhir/ -c output/true.cql -u http://localhost:8080/cqf-ruler/baseDstu3/fhir
> ```
> 
> and got this error:
> 
> ```
> Posting bundle /Users/tstrassner/repo/synthea/output/fhir/Bart73_Hickle134_6c89018f-09dc-419b-922d-aa65d73089e2.json
> (node:58090) UnhandledPromiseRejectionWarning: Error: Failed to post bundle
>     at processBundles (/Users/tstrassner/repo/fhir-bundle-calculator/src/cli.js:85:13)
>     at <anonymous>
>     at process._tickCallback (internal/process/next_tick.js:189:7)
> (node:58090) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
> (node:58090) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> ```

@tstrass looks like the wrong cqf ruler base url. I was using `/cqf-ruler-dstu3/fhir`. I don't think they use `baseDstu3` anymore

File: src/cli.js [hunk 1]
```diff
@@ -1,8 +1,134 @@
...
+}
+
+// Create output directory if it doesn't exist
+const outputRoot = './output';
+if (!fs.existsSync(outputRoot)) {
+  fs.mkdirSync(outputRoot);
+}
+
+// Output for this run is timestamped with the current datetime
+const outputPath = path.join(outputRoot, `/results-${moment().format()}`);
```
[the remark this answers is not in the document]
Comment (mgramigna) [reply src/cli.js]: Good point I'll format it differently

**Comment by tstrass** [comment]
I tried that too, but same result. FWIW it says this in the CQF Ruler web UI.
Screen Shot 2019-11-14 at 1 17 54 PM

**Comment by mgramigna** [comment]
If that is your base, then don't include `/fhir`

**Comment by mgramigna** [comment]
I just pushed up a change that prints a basic error message when it fails to post the bundle. Hopefully now your other case should result in 404 because the base was wrong?

I should've had this in before. Might've avoided this back and forth

**Comment by tstrass** [comment]
I'm super unfamiliar with how you're supposed to manage a package.json, but running an `npm install` resulted in `handlebars` version `4.5.2` for me instead of `4.5.1`. Should we update that in the package-lock.json?

**Comment by mgramigna** [comment]
`package-lock.json` is autogenerated and should not be manually edited. I'm not sure why that's happening, but handlebars isn't an explicit dependency of this library, it's probably a dependency of a dependency of a dependency or something. I'm not too concerned about it

**Comment by mgramigna** [comment]
I regenerated a clean package-lock. I think I had something local that wasn't checked in

**Comment by tstrass** [comment]
Ok, the only thing i get differently in the package-lock now is that it adds `"optional": true` to a bunch of packages. Just letting you know in case that's unexpected, not requesting a change.

**Comment by mgramigna** [comment]
Yeah good call with the duplication stuff. Wasn't sure about that. Will look into it now

**Comment by mgramigna** [comment]
Just pushed up a change to only sort bundles into their deepest drilled-down subset. Updated docs accordingly

File: src/utils/calculation.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *   numerator: true|false,
+ *   denominator: true|false
+ * }
+ */
+const calculate = async (url, cql, patientId, periodStart, periodEnd) => {
+  const parameters = buildParameters(cql, patientId, periodStart, periodEnd);
+  const response = await axios.post(`${url}/$cql`, parameters);
+
+  // Get results for only the definitions that we care about
+  const definitions = ['Initial Population', 'Numerator', 'Denominator'];
```
Comment (mayerm94) [inline src/utils/calculation.js]: I would add 'Error' to this array and then logic in cli.js to write to an error directory (see other comment)

File: src/cli.js [hunk 3]
```diff
@@ -1,8 +1,131 @@
...
+    if (res.numerator) {
+      console.log(`Wrote bundle ${bundlePath} to ${numerPath}`);
+      fs.writeFileSync(`${numerPath}/${path.basename(bundlePath)}`, bundleContent, 'utf8');
+    } else if (res.denominator) {
+      console.log(`Wrote bundle ${bundlePath} to ${denomPath}`);
+      fs.writeFileSync(`${denomPath}/${path.basename(bundlePath)}`, bundleContent, 'utf8');
+    } else if (res.initial_population) {
+      console.log(`Wrote bundle ${bundlePath} to ${ipopPath}`);
+      fs.writeFileSync(`${ipopPath}/${path.basename(bundlePath)}`, bundleContent, 'utf8');
+    }
```
Comment (mayerm94) [inline src/cli.js]: Would add res.error here and write to error directory.

**Comment by mgramigna** [comment]
@mayerm94 pushed up some error handling changes

**Pull request merged**